src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -579,6 +599,34 @@ template <typename DataType, typename DistType>
 int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType label,
                                                    void *auxiliaryCtx) {
     int ret = 1;
+    auto hnsw_index = this->getHNSWIndex();
+    if (this->getWriteMode() == VecSim_WriteInPlace) {
+        this->mainIndexGuard.lock();
+        // Internally, we may overwrite (delete the previous vector stored under this label), and
+        // may need to increase the capacity when we append the new vector afterwards.
+        ret = hnsw_index->addVector(blob, label);
+        this->mainIndexGuard.unlock();
+        this->UpdateIndexMemory(this->memoryCtx, this->getAllocationSize());
+        return ret;
+    }
+    if (this->frontendIndex->indexSize() >= this->flatBufferLimit) {
```

Comment:
Don't you need to lock here?

Reply:
I don't think so... see https://github.com/RedisAI/VectorSimilarity/pull/361#discussion_r1181255582